video/cursor.h

```diff
@@ -43,34 +59,116 @@ inline void setActiveCursor(Point * cursor) {
 }
 
 inline void setCursorBehaviour(uint8_t setting, uint8_t mask = 0xFF) {
-	cursorBehaviour = (cursorBehaviour & mask) ^ setting;
+	cursorBehaviour.value = (cursorBehaviour.value & mask) ^ setting;
 }
 
 // Move the active cursor to the leftmost position in the viewport
 //
 void cursorCR() {
-	activeCursor->X = activeViewport->X1;
+	if (cursorBehaviour.flipXY) {
+		activeCursor->Y = cursorBehaviour.invertVertical ? (activeViewport->Y2 + 1 - fontH) : activeViewport->Y1;
+	} else {
+		activeCursor->X = cursorBehaviour.invertHorizontal ? (activeViewport->X2 + 1 - fontW) : activeViewport->X1;
+	}
+}
+
+bool cursorIsOffRight() {
+	if (cursorBehaviour.flipXY) {
+		if (cursorBehaviour.invertHorizontal) {
+			return activeCursor->Y < activeViewport->Y1;
+		}
+		return activeCursor->Y >= activeViewport->Y2;
+	}
+	if (cursorBehaviour.invertHorizontal) {
+		return activeCursor->X < activeViewport->X1;
+	}
+	return activeCursor->X >= activeViewport->X2;
+}
+
+bool cursorScrollOrWrap() {
+	bool outsideY = activeCursor->Y < activeViewport->Y1 || activeCursor->Y >= activeViewport->Y2;
+	bool outsideX = activeCursor->X < activeViewport->X1 || activeCursor->X >= activeViewport->X2;
+	if (!outsideX && !outsideY) {
+		// cursor within current viewport, so do nothing
+		return false;
+	}
+
+	if (textCursorActive() && !cursorBehaviour.yWrap) {
+		// text cursor, scrolling for our Y direction is enabled
+		if (cursorBehaviour.flipXY && outsideX || !cursorBehaviour.flipXY && outsideY) {
+			// we are outside of our cursor vertical direction
+			// scroll in appropriate direction by 1 character (movement amount 0)
+			// our direction will be a 6 or 7, depending on whether we're outside the "top" or "bottom"
+			if (outsideX) {
+				if (activeCursor->X < activeViewport->X1) {
+					scrollRegion(activeViewport, cursorBehaviour.invertVertical ? 7 : 6, 0);
+					activeCursor->X = activeViewport->X1;
+				} else {
+					scrollRegion(activeViewport, cursorBehaviour.invertVertical ? 6 : 7, 0);
+					activeCursor->X = activeViewport->X2 + 1 - fontW;
+				}
+			} else {
+				if (activeCursor->Y < activeViewport->Y1) {
+					scrollRegion(activeViewport, cursorBehaviour.invertHorizontal ? 7 : 6, 0);
+					activeCursor->Y = activeViewport->Y1;
+				} else {
+					scrollRegion(activeViewport, cursorBehaviour.invertHorizontal ? 6 : 7, 0);
+					activeCursor->Y = activeViewport->Y2 + 1 - fontH;
+				}
+			}
+			return false;
+		}
+	}
+
+	// if we get here we have a graphics cursor, or text cursor with wrap enabled
+
+	if (!textCursorActive() && cursorBehaviour.grNoSpecialActions) {
+		return false;
+	}
+
+	// here we just wrap everything
+	// TODO these don't work properly if our viewport isn't a multiple of fontW or fontH
+	// NB this will only be a potential problem if we're using a graphics cursor with a restricted viewport
+	// or potentially if we're using an oddly sized font
```

Comment:
as noted here, these cursor wrapping movements won't necessarily work properly when a viewport is set that is not a clean integer multiple of the font width or height

practically speaking, right now this is only a problem when using the graphics cursor, and for general text printing shouldn't be a significant issue.  it essentially only is an issue when using text at the graphics cursor for interactive editing going beyond the end of the line, which has a whole bunch of other subtle issues associated with it anyway.

when we add support for different font sizes this will also affect text cursor so at that point we should revisit this and ensure that we have a better solution